File: 0486_Predict the Winner/PredicttheWinner_DP.cs
public class Solution
{
    public bool PredictTheWinner(int[] nums)
    {
        var n = nums.Length;
        var dp = new int[n][];

        for(int i=0;i<dp.Length;i++)
        {
            dp[i] = new int[n];
            Array.Fill(dp[i], int.MinValue);
        }

        for (var i = 0; i < n; i++)
            dp[i][i] = nums[i];

        for (var len = 2; len <= n; len++)
        {
            for (var i = 0; i <= n - len; i++)
            {
                var j = i + len - 1;
                dp[i][j] = Math.Max(nums[i] - dp[i + 1][j], nums[j] - dp[i][j - 1]);
            }
        }

        return dp[0][n - 1] >= 0;
    }
}

File: 0486_Predict the Winner/PredicttheWinner_DP.java

class Solution {
    public boolean PredictTheWinner(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n][n];
        for (int[] rows: dp)
            Arrays.fill(rows, Integer.MIN_VALUE);
        
        for(int i=0;i<n;i++)
            dp[i][i] = nums[i];
        
        for(int len=2;len<=n;len++){
            for(int l=0; l<=n-len;l++){
                int r=l+len-1;
                dp[l][r] = Math.max(nums[l] - dp[l + 1][r], nums[r] - dp[l][r - 1]);
            }
        }
    return dp[0][n-1] >= 0;
    }
}